storage: fix Raft log size accounting

tbg opened the pull request:
We were accounting for sideloaded payloads (SSTs) when adding them to
the log, but were omitting them during truncations. As a result, the
tracked raft log size would permanently skyrocket which in turn would
lead to extremely aggressive truncations and resulted in pathological
amounts of Raft snapshots.

I'm still concerned about this logic as we're now relying on numbers
obtained from the file system to match exactly a prior in-mem
computation, and there may be other bugs that cause a divergence. But
this fixes the blatant obvious one, so it's a step in the right
direction.

The added tests highlight a likely omission in the sideloaded storage
code which doesn't access the file system through the RocksDB env as it
seems like it should, filed as #31913.

At this point it's unclear whether it fixes the below issues, but at the
very least it seems to address a major problem they encountered:

Touches #31732.
Touches #31740.
Touches #30261.
Touches #31768.
Touches #31745.

Release note (bug fix): avoid a performance degradation related to
overly aggressive Raft log truncations that could occur during RESTORE
or IMPORT operations.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/31914)
<!-- Reviewable:end -->

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Thanks for pushing this over the finish line.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31914#-:-LPkWA1P13C_-XmNeBzm:bv170aj)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/cmd/roachtest/tpcc.go, line 362 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31914#-LPkV8BC0dbqHOoYlNur:-LPkV8BC0dbqHOoYlNus:b-d6z0fw) ([raw file](https://github.com/cockroachdb/cockroach/blob/51f40c5265fc32bfced692fdf6cc8b9c7fb3e25d/pkg/cmd/roachtest/tpcc.go#L362)):*
> ```Go
> 	if _, err := db.ExecContext(ctx, `SET CLUSTER SETTING kv.range_merge.queue_enabled = false`); err != nil {
> 		return err
> 	}
> ```

Perhaps revert the changes in this file which I made to debug tpcc-bench stuff.

@benesch might have an opinion about disabling the merge queue for tpcc bench runs until the merge issues are sorted out.
___
*[pkg/storage/replica.go, line 6584 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31914#-LPkVMrVD3LZ-xtOZoct:-LPkVMrVD3LZ-xtOZocu:brcj0t5) ([raw file](https://github.com/cockroachdb/cockroach/blob/51f40c5265fc32bfced692fdf6cc8b9c7fb3e25d/pkg/storage/replica.go#L6584)):*
> ```Go
> 	// Gossip the cluster ID from all replicas of the first range; there
> 	// is no expiration on the cluster ID.
> 	if false && log.V(1) {
> ```

There are a bunch of log changes such as this that should be reverted.
___
*[pkg/storage/replica.go, line 6617 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31914#-LPkVRlJ8NTSnR_oUIXr:-LPkVRlJ8NTSnR_oUIXs:ba854ie) ([raw file](https://github.com/cockroachdb/cockroach/blob/51f40c5265fc32bfced692fdf6cc8b9c7fb3e25d/pkg/storage/replica.go#L6617)):*
> ```Go
> 	log.Event(ctx, "gossiping sentinel and first range")
> 	if log.V(1) {
> 		//log.Infof(ctx, "gossiping sentinel from store %d, r%d", r.store.StoreID(), r.RangeID)
> ```

Revert?
___
*[pkg/storage/replica.go, line 6624 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31914#-LPkVSjm8DaDFAp4rOBA:-LPkVSjm8DaDFAp4rOBB:ba854ie) ([raw file](https://github.com/cockroachdb/cockroach/blob/51f40c5265fc32bfced692fdf6cc8b9c7fb3e25d/pkg/storage/replica.go#L6624)):*
> ```Go
> 		log.Errorf(ctx, "failed to gossip sentinel: %s", err)
> 	}
> 	if false && log.V(1) {
> ```

Revert?
___
*[pkg/storage/replica_proposal.go, line 573 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31914#-LPkVVPV4wvQpIvcLIaB:-LPkVVPV4wvQpIvcLIaC:b-bo6h8j) ([raw file](https://github.com/cockroachdb/cockroach/blob/51f40c5265fc32bfced692fdf6cc8b9c7fb3e25d/pkg/storage/replica_proposal.go#L573)):*
> ```Go
> 					log.Errorf(ctx, "while removing sideloaded files during log truncation: %s", err)
> 				} else {
> 					rResult.RaftLogDelta -= size
> ```

I trust that you agree this is copacetic. This was the part of the change that I was most unsure about.


<!-- Sent from Reviewable.io -->

tschottdorf commented:
I ran a few roachtests to verify that things are stable now.

These two look good (another two running now):

```
--- PASS: restore2TB/nodes=10 (8959.51s)
--- PASS: restore2TB/nodes=10 (8743.44s)
```

But I got this suspicious failure:

```
--- FAIL: import/tpch/nodes=8 (9943.52s)
	test.go:639,cluster.go:1453,import.go:122: pq: split at key /Table/53/1/390157478/5 failed: replica 0 is invalid: ReplicaID must not be zero
```

This originates in https://github.com/cockroachdb/cockroach/blob/a050d569f236fee50960e9b45f3052a6e0d364b4/pkg/roachpb/metadata.go#L222-L234

So we have a replica that has a NodeID and StoreID but a zero replicaID either in https://github.com/cockroachdb/cockroach/blob/62ea444e7b393bbd27e30705e1b04451bd21f49b/pkg/storage/replica_command.go#L348

or in the other call to `updateRangeDescriptor`. I'm taking a look at the logs now. Doubtful that it's at all related to this change.

PS merges are turned off in this testing.

tschottdorf commented:
Filed that error as a separate issue https://github.com/cockroachdb/cockroach/issues/31918.

tbg committed:
storage: fix Raft log size accounting

We were accounting for sideloaded payloads (SSTs) when adding them to
the log, but were omitting them during truncations. As a result, the
tracked raft log size would permanently skyrocket which in turn would
lead to extremely aggressive truncations and resulted in pathological
amounts of Raft snapshots.

I'm still concerned about this logic as we're now relying on numbers
obtained from the file system to match exactly a prior in-mem
computation, and there may be other bugs that cause a divergence. But
this fixes the blatant obvious one, so it's a step in the right
direction.

The added tests highlight a likely omission in the sideloaded storage
code which doesn't access the file system through the RocksDB env as it
seems like it should, filed as #31913.

At this point it's unclear whether it fixes the below issues, but at the
very least it seems to address a major problem they encountered:

Touches #31732.
Touches #31740.
Touches #30261.
Touches #31768.
Touches #31745.

Release note (bug fix): avoid a performance degradation related to
overly aggressive Raft log truncations that could occur during RESTORE
or IMPORT operations.

tbg reviewed (commented):
> Reviewed 5 of 9 files at r1, 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31914)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/replica_proposal.go, line 573 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31914#-LPkVVPV4wvQpIvcLIaB:-LPk_e2K9btvG9UA3qAd:b-btgxxt) ([raw file](https://github.com/cockroachdb/cockroach/blob/51f40c5265fc32bfced692fdf6cc8b9c7fb3e25d/pkg/storage/replica_proposal.go#L573)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I trust that you agree this is copacetic. This was the part of the change that I was most unsure about.
</blockquote></details>

Are you concerned about anything in particular? This seems fine to me. We could compute this upstream, but it doesn't seem better, and we don't need this to be equivalent between replicas anyway.
One problem could be that we might be removing more than we added. For example, if a bunch of data gets sideloaded and then the cluster restarts, the first truncation after the restart will get lots of bytes here "for free". This is likely going to result in a raft log size of zero, and it would be a little off (too small) from that point on. (This isn't a problem that needs sideloaded payloads, it's basically by design). This can only result in a de facto size that is off by a factor of two (i.e. truncating 2x as late as configured), so it's probably not a big deal.
It'll be painful to get all of this "correct". We can't simply recompute the size of the remaining log here as there might be more sideloaded payloads. Besides, sideloaded payloads for older terms may coexist with new ones.

I just wrote a comment to deal with residual space in Raft logs for idle ranges here: https://github.com/cockroachdb/cockroach/blob/e808caf5eb5c70d0d4072bcef94843575cfe66ea/pkg/storage/raft_log_queue.go#L38-L41
Perhaps when performing such a truncation, we could do a little more work to recompute the exact length. I hope it's never going to be worth it though.
___
*[pkg/cmd/roachtest/tpcc.go, line 362 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31914#-LPkV8BC0dbqHOoYlNur:-LPka7dqB8IMi5qCXfqd:b-avx650) ([raw file](https://github.com/cockroachdb/cockroach/blob/51f40c5265fc32bfced692fdf6cc8b9c7fb3e25d/pkg/cmd/roachtest/tpcc.go#L362)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Perhaps revert the changes in this file which I made to debug tpcc-bench stuff.

@benesch might have an opinion about disabling the merge queue for tpcc bench runs until the merge issues are sorted out.
</blockquote></details>

Not intentional. Sorry for not catching that.
___
*[pkg/storage/replica.go, line 6584 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31914#-LPkVMrVD3LZ-xtOZoct:-LPkaAztE_-Y6vLKba4L:b-avx650) ([raw file](https://github.com/cockroachdb/cockroach/blob/51f40c5265fc32bfced692fdf6cc8b9c7fb3e25d/pkg/storage/replica.go#L6584)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

There are a bunch of log changes such as this that should be reverted.
</blockquote></details>

Not intentional. Sorry for not catching that.
___
*[pkg/storage/replica.go, line 6617 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31914#-LPkVRlJ8NTSnR_oUIXr:-LPkaBOwCpHGs8_x0oZ-:b-avx650) ([raw file](https://github.com/cockroachdb/cockroach/blob/51f40c5265fc32bfced692fdf6cc8b9c7fb3e25d/pkg/storage/replica.go#L6617)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Revert?
</blockquote></details>

Not intentional. Sorry for not catching that.
___
*[pkg/storage/replica.go, line 6624 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31914#-LPkVSjm8DaDFAp4rOBA:-LPkaCDX81Ev52Vm7gQ5:b-avx650) ([raw file](https://github.com/cockroachdb/cockroach/blob/51f40c5265fc32bfced692fdf6cc8b9c7fb3e25d/pkg/storage/replica.go#L6624)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Revert?
</blockquote></details>

Not intentional. Sorry for not catching that.


<!-- Sent from Reviewable.io -->

petermattis reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31914)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/replica_proposal.go, line 573 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31914#-LPkVVPV4wvQpIvcLIaB:-LPklDVb0jz8i4FjuGGT:b16bif1) ([raw file](https://github.com/cockroachdb/cockroach/blob/51f40c5265fc32bfced692fdf6cc8b9c7fb3e25d/pkg/storage/replica_proposal.go#L573)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Are you concerned about anything in particular? This seems fine to me. We could compute this upstream, but it doesn't seem better, and we don't need this to be equivalent between replicas anyway.
One problem could be that we might be removing more than we added. For example, if a bunch of data gets sideloaded and then the cluster restarts, the first truncation after the restart will get lots of bytes here "for free". This is likely going to result in a raft log size of zero, and it would be a little off (too small) from that point on. (This isn't a problem that needs sideloaded payloads, it's basically by design). This can only result in a de facto size that is off by a factor of two (i.e. truncating 2x as late as configured), so it's probably not a big deal.
It'll be painful to get all of this "correct". We can't simply recompute the size of the remaining log here as there might be more sideloaded payloads. Besides, sideloaded payloads for older terms may coexist with new ones.

I just wrote a comment to deal with residual space in Raft logs for idle ranges here: https://github.com/cockroachdb/cockroach/blob/e808caf5eb5c70d0d4072bcef94843575cfe66ea/pkg/storage/raft_log_queue.go#L38-L41
Perhaps when performing such a truncation, we could do a little more work to recompute the exact length. I hope it's never going to be worth it though.
</blockquote></details>

No particular concerns, but this is the only place where we're computing `RaftLogDelta` below Raft.


<!-- Sent from Reviewable.io -->

tschottdorf commented:
Ack. I'm going to merge but would like to hear if @nvanbenschoten has any concerns and would rather move this upstream of Raft.

bors r=petermattis

nvb reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> I don't think this can be moved upstream into Raft because `etcd/raft` has no concept of log truncation. My understanding is that it assumes that logs are persisted forever.

> Reviewed 4 of 9 files at r1, 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31914#-:-LPkub_LBamFhzw7rv37:bfd949w)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale)

___
*[pkg/storage/replica_proposal.go, line 573 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31914#-LPkVVPV4wvQpIvcLIaB:-LPktZY-9OrepXMb0MgM:bgep1tq) ([raw file](https://github.com/cockroachdb/cockroach/blob/51f40c5265fc32bfced692fdf6cc8b9c7fb3e25d/pkg/storage/replica_proposal.go#L573)):*
> Besides, sideloaded payloads for older terms may coexist with new ones.

That's interesting. Don't we run into the same issue that we do here when a new term overwrites old sideloaded entries? Should we be doing something similar to this here:

https://github.com/cockroachdb/cockroach/blob/0368172b6fec0c91949ea7a3f1ef706acd7eae66/pkg/storage/replica.go#L4333-L4340


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
Sorry, I didn't mean "into etcd/raft" but "into proposal evaluation".

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31914#-:-LPkxXGwByEI6-uHa_AR:b-1r40q1)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/replica_proposal.go, line 573 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31914#-LPkVVPV4wvQpIvcLIaB:-LPkxmN0DI343DUvXz8B:b-d1rg1q) ([raw file](https://github.com/cockroachdb/cockroach/blob/51f40c5265fc32bfced692fdf6cc8b9c7fb3e25d/pkg/storage/replica_proposal.go#L573)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> Besides, sideloaded payloads for older terms may coexist with new ones.

That's interesting. Don't we run into the same issue that we do here when a new term overwrites old sideloaded entries? Should we be doing something similar to this here:

https://github.com/cockroachdb/cockroach/blob/0368172b6fec0c91949ea7a3f1ef706acd7eae66/pkg/storage/replica.go#L4333-L4340
</blockquote></details>

Absolutely, let me look into that.


<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
> Sorry, I didn't mean "into etcd/raft" but "into proposal evaluation".

Ah, I see. I don't have any big concerns about what we do here. If computing this during proposal evaluation means that we would need to peek at each Raft log entry (in Go) then I'd say it's a non-starter. If we could just look directly at the sideloaded storage, that might be a workable solution. It would be nice to avoid looking at the file size on each replica beneath Raft and hoping they're all the same, even if it is ok for them to diverge. It would also be nice to keep this out of the single-threaded Raft loop.
